File: server/src/main/java/org/example/server/model/Project.java
package org.example.server.model;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.util.List;

@Entity
@EntityListeners(AuditingEntityListener.class)
@Table(name = "project")
@Getter
@Setter
@Builder
@ToString(exclude = {"userProjects", "createdBy", "columns"})
@EqualsAndHashCode(of = "id")
@NoArgsConstructor
@AllArgsConstructor
public class Project {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 50)
    private String name;

    @Column(length = 100)
    private String description;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by", nullable = false)
    @JsonManagedReference
    private User createdBy;

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<UserProject> userProjects;

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<BoardColumn> columns;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDate createdDate;

    @LastModifiedDate
    @Column
    private LocalDate updatedDate;

}

File: client/src/types/interfaces/project.ts
import type { BoardColumnResponse } from "./boardColumn";
import type { UserSimplifiedResponse } from "./userSimplified";

export interface Project {
  name: string;
  description?: string;
  startDate: string;
  endDate: string;
}

export interface ProjectRequest {
  name: string;
  description?: string;
  startDate: string;
  endDate: string;
  createdBy: number;
}

export interface ProjectResponse {
  id: number;
  name: string;
  description?: string;
  startDate: string;
  endDate: string;
  createdDate: string;
  updatedDate?: string;
  createdBy: UserSimplifiedResponse;
  users: UserSimplifiedResponse[];
  columns: BoardColumnResponse[];
}

